### ISSUE TITLE
Consider using Lua's coroutine for functions where just.doUntil has been used

### ISSUE THREAD

**Issue by cailyoung** [fcpxhacks/fcpxhacks#963]
As CommandPost shares the UI thread with every other function, any use of `just.doUntil` will block the UI (if I'm not mistaken).

I posted on another issue about this but thought it was worth raising as an enhancement request - yes, some of the usages of `just.doUntil` will be fine but I imagine some of them should probably leverage `coroutine` to avoid blocking the UI.

**Comment by latenitefilms** [comment]
Coroutines are above my knowledge/skillset - thoughts @randomeizer ?

**Comment by randomeizer** [comment]
Yeah, they definitely block the thread. That was intentional in the design,
because we needed to "wait" for the UI (typically) to update, and using a
timer to offload them would complicate other functions that were expecting
it to compete when it returned.

Open to suggestions on improving it.

**Comment by cailyoung** [comment]
@randomeizer I guess this would only be of value for time-consuming processes that the user doesn't need to wait on a UI update for, and even then only ones likely to take more than 0.2s

The FCP reload was the example that got me looking but I imagine some of the startup plugin scanning etc could benefit as well?

Having read in to it a bit more, they aren't actually parallel threads but do allow rapid context switching so you could have the (relatively) long-running process yield back to the UI every 0.02s or something.

**Comment by randomeizer** [comment]
Are coroutines the same as timers?

For things like app reloads, they could be useful. We also have the ability
to watch for various app events, which may be more efficient.

**Comment by latenitefilms** [comment]
See: 

- [9.1 – Coroutine Basics](https://www.lua.org/pil/9.1.html)
- [Coroutines Tutorial](https://github.com/CommandPost/CommandPost/issues/963)
- [A Field Guide to Lua Coroutines](https://spin.atomicobject.com/2013/07/01/lua-coroutines/)

**Comment by randomeizer** [comment]
Coroutines are a bit different to timers. They essentially allow a way of
letting a loop relinquish control periodically, and then pick it back up
again later. We couldn't just drop them in for our "just" operations, but
they are certainly something that would be more useful if/when we start
moving to RxLua/Observable APIs. RxLua has some built in support for
coroutines actually.

Regards,

David


On 7 Jun 2018 at 10:10 pm, Chris Hocking <notifications@github.com> wrote:


See:

   - 9.1 – Coroutine Basics <https://www.lua.org/pil/9.1.html>
   - Coroutines Tutorial
   <https://github.com/CommandPost/CommandPost/issues/963>
   - A Field Guide to Lua Coroutines
   <https://spin.atomicobject.com/2013/07/01/lua-coroutines/>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/CommandPost/CommandPost/issues/963#issuecomment-395397944>,
or mute the thread

**Comment by latenitefilms** [comment]
All sounds good - I'm going to close this issue for now then.
